examples/mxnet/sampling/dis_sampling/gcn_trainer.py

```diff
@@ -0,0 +1,269 @@
+                             fn.copy_src(src='h', out='m'),
+                             fn.sum(msg='m', out='h'),
+                             layer)
+
+        h = nf.layers[-1].data.pop('activation')
+        return h
+
+
+def main(args):
+    # load and preprocess dataset
+    data = load_data(args)
+
+    if args.gpu >= 0:
+        ctx = mx.gpu(args.gpu)
+    else:
+        ctx = mx.cpu()
+
+    if args.self_loop and not args.dataset.startswith('reddit'):
+        data.graph.add_edges_from([(i,i) for i in range(len(data.graph))])
+
+    # Create sampler receiver
+    receiver = dgl.contrib.sampling.SamplerReceiver(ip=args.ip, port=args.port, num_sender=args.num_sender)
+
+    train_nid = mx.nd.array(np.nonzero(data.train_mask)[0]).astype(np.int64).as_in_context(ctx)
+    test_nid = mx.nd.array(np.nonzero(data.test_mask)[0]).astype(np.int64).as_in_context(ctx)
+
+    features = mx.nd.array(data.features).as_in_context(ctx)
+    labels = mx.nd.array(data.labels).as_in_context(ctx)
+    train_mask = mx.nd.array(data.train_mask).as_in_context(ctx)
+    val_mask = mx.nd.array(data.val_mask).as_in_context(ctx)
+    test_mask = mx.nd.array(data.test_mask).as_in_context(ctx)
+    in_feats = features.shape[1]
+    n_classes = data.num_labels
+    n_edges = data.graph.number_of_edges()
+
+    n_train_samples = train_mask.sum().asscalar()
+    n_val_samples = val_mask.sum().asscalar()
+    n_test_samples = test_mask.sum().asscalar()
+
+    print("""----Data statistics------'
+      #Edges %d
+      #Classes %d
+      #Train samples %d
+      #Val samples %d
+      #Test samples %d""" %
+          (n_edges, n_classes,
+              n_train_samples,
+              n_val_samples,
+              n_test_samples))
+
+    # create GCN model
+    g = DGLGraph(data.graph, readonly=True)
+
+    g.ndata['features'] = features
+
+    num_neighbors = args.num_neighbors
+
+    degs = g.in_degrees().astype('float32').as_in_context(ctx)
+    norm = mx.nd.expand_dims(1./degs, 1)
+    g.ndata['norm'] = norm
+
+    model = GCNSampling(in_feats,
+                        args.n_hidden,
+                        n_classes,
+                        args.n_layers,
+                        mx.nd.relu,
+                        args.dropout,
+                        prefix='GCN')
+
+    model.initialize(ctx=ctx)
+    loss_fcn = gluon.loss.SoftmaxCELoss()
+
+    infer_model = GCNInfer(in_feats,
+                           args.n_hidden,
+                           n_classes,
+                           args.n_layers,
+                           mx.nd.relu,
+                           prefix='GCN')
+
+    infer_model.initialize(ctx=ctx)
+
+    # use optimizer
+    print(model.collect_params())
+    trainer = gluon.Trainer(model.collect_params(), 'adam',
+                            {'learning_rate': args.lr, 'wd': args.weight_decay},
+                            kvstore=mx.kv.create('local'))
+
+    # initialize graph
+    dur = []
+    total_count = 153
```

Comment:
it is very difficult for users to figure what is the right number of batches in an epoch.

Reply:
Yes it is. Actually I don't know how to solve this problem easily for now.

Reply:
The potential solution could be:

1. Send a number to trainer from sender to tell trainer what the this number is.
2. Batch send all the sub-graph in an epoch to trainer. But it may cause the memory and network issue because this batched-nodeflow could be extremely large.

Reply:
I think we can just leave this problem here, and solve it after all of the distributed training pipeline completed.

Reply:
Why do we want the number of batches here?  I don't see why this is a big deal...

Reply:
We need to deliver runnable code to users before 0.3 release.

To really fix this problem, I think we should provide users the notion of iterator and epoch. Just like normal samplers.

Reply:
what do you think of this API?
```python
receiver = dgl.contrib.sampling.SamplerReceiver(ip=args.ip, port=args.port, num_sender=args.num_sender)
for epoch in range(n_epochs):
    for subg in receiver.get_sampler():
        ...
```

Reply:
Maybe we propose a set of interface for distributed sampler and have everyone involved in discussion, just like what we did for all other interface before?

Reply:
This API is good. I think we should figure out a way to define the number of nodeflow in an epoch when all the sampler processes's out are mixed.

Reply:
A navie way to solve this problem is that we let one sampler pre-sample an epoch, and send the size of sampled nodeflow to trainer node, and we can use this number to define the boundary of epoch. But it is not very friendly for users who write code in sampler.